File: packages/sdk/src/integrations/autogen.ts
/**
 * AgentGuard — AutoGen Integration
 *
 * Provides guard wrappers for Microsoft's AutoGen framework that intercept
 * tool execution and evaluate each call through AgentGuard's policy engine.
 *
 * Two integration styles:
 *
 * 1. **Function wrapper** — wrap individual tool functions:
 *    ```typescript
 *    import { createAutoGenGuard } from '@the-bot-club/agentguard';
 *
 *    const guard = createAutoGenGuard({ apiKey: 'ag_...' });
 *    const guardedTool = guard.wrapTool('read_file', readFileFn);
 *    ```
 *
 * 2. **Class-based middleware** — intercept all tool calls on an agent:
 *    ```typescript
 *    import { AutoGenToolGuard } from '@the-bot-club/agentguard';
 *
 *    const guard = new AutoGenToolGuard({ apiKey: 'ag_...' });
 *    guard.patchAgent(conversableAgent);
 *    ```
 *
 * Both approaches use duck-typing — no hard dependency on `autogen` or
 * `@microsoft/autogen`. Works with any structurally compatible agent.
 */
import { AgentGuard } from '../sdk/client.js';
import { AgentGuardBlockError } from './errors.js';

// ─── Options ──────────────────────────────────────────────────────────────────

export interface AutoGenGuardOptions {
  /** AgentGuard API key (ag_...) */
  apiKey: string;
  /** Override the AgentGuard API base URL */
  baseUrl?: string;
  /** Optional agent ID for scoped evaluations */
  agentId?: string;
  /**
   * When true, throw AgentGuardBlockError if a tool call is blocked.
   * When false (default), call onBlock and return undefined (swallow the block).
   */
  throwOnBlock?: boolean;
  /** Callback invoked when a tool call is blocked. */
  onBlock?: (info: AutoGenBlockInfo) => void;
  /** Callback invoked when a tool call is allowed. */
  onAllow?: (info: AutoGenAllowInfo) => void;
}

// ─── Callback info types ──────────────────────────────────────────────────────

export interface AutoGenBlockInfo {
  tool: string;
  params: Record<string, unknown>;
  decision: string;
  riskScore: number;
  reason?: string;
  suggestion?: string;
  alternatives?: string[];
  matchedRuleId?: string;
}

export interface AutoGenAllowInfo {
  tool: string;
  params: Record<string, unknown>;
  decision: 'allow' | 'monitor';
  riskScore: number;
  matchedRuleId?: string;
}

// ─── Minimal AutoGen structural types (duck-typed, no hard dep) ───────────────

/** Any function that can serve as an AutoGen tool. */
type ToolFunction = (...args: unknown[]) => unknown;

/**
 * Structural interface for AutoGen's ConversableAgent (duck-typed).
 * AutoGen agents that register and execute tools expose these members.
 */
interface ConversableAgentLike {
  /** Registered tool functions keyed by name. */
  _tools?: Record<string, ToolFunction>;
  /** Alternative: tool map used in some AutoGen versions. */
  registered_tools?: Record<string, ToolFunction>;
  /** Tool execution method (AutoGen v0.2+). */
  execute_tool?: (toolName: string, args: Record<string, unknown>) => Promise<unknown>;
  /** Tool execution method (AutoGen v0.4+). */
  execute_function?: (toolName: string, args: Record<string, unknown>) => Promise<unknown>;
  [key: string]: unknown;
}

// ─── Internal helpers ─────────────────────────────────────────────────────────

function normaliseDecision(raw: Record<string, unknown>): {
  decision: string;
  riskScore: number;
  reason?: string;
  suggestion?: string;
  alternatives?: string[];
  matchedRuleId?: string;
} {
  return {
    decision: (raw['result'] ?? raw['decision'] ?? 'allow') as string,
    riskScore: (raw['riskScore'] as number | undefined) ?? 0,
    reason: raw['reason'] as string | undefined,
    suggestion: raw['suggestion'] as string | undefined,
    alternatives: raw['alternatives'] as string[] | undefined,
    matchedRuleId: raw['matchedRuleId'] as string | undefined,
  };
}

// ─── AutoGenToolGuard class ───────────────────────────────────────────────────

/**
 * Class-based guard that can be used as middleware with AutoGen's
 * ConversableAgent. Patches an agent's tool execution to evaluate
 * every call through AgentGuard before allowing it to proceed.
 *
 * @example
 * ```typescript
 * import { AutoGenToolGuard } from '@the-bot-club/agentguard';
 *
 * const guard = new AutoGenToolGuard({ apiKey: 'ag_...' });
 *
 * // Patch an existing agent
 * guard.patchAgent(myAgent);
 *
 * // Or evaluate manually before execution
 * const result = await guard.evaluate('send_email', { to: 'user@example.com' });
 * if (result.blocked) {
 *   console.log('Blocked:', result.reason);
 * }
 * ```
 */
export class AutoGenToolGuard {
  private readonly guard: AgentGuard;
  private readonly options: AutoGenGuardOptions;
  private readonly throwOnBlock: boolean;

  constructor(options: AutoGenGuardOptions) {
    this.options = options;
    this.throwOnBlock = options.throwOnBlock ?? false;
    this.guard = new AgentGuard({
      apiKey: options.apiKey,
      baseUrl: options.baseUrl,
    });
  }

  /**
   * Evaluate a tool call against the policy engine.
   *
   * @param toolName  Name of the tool
   * @param params    Tool arguments
   * @returns         Evaluation result with blocked flag
   */
  async evaluate(
    toolName: string,
    params: Record<string, unknown> = {},
  ): Promise<{
    blocked: boolean;
    decision: string;
    riskScore: number;
    reason?: string;
    suggestion?: string;
    alternatives?: string[];
    matchedRuleId?: string;
  }> {
    const result = await this.guard.evaluate({ tool: toolName, params });
    const normalised = normaliseDecision(result as unknown as Record<string, unknown>);
    const blocked = normalised.decision === 'block' || normalised.decision === 'require_approval';

    if (blocked) {
      this.options.onBlock?.({
        tool: toolName,
        params,
        ...normalised,
      });

      if (this.throwOnBlock) {
        throw new AgentGuardBlockError({
          ...normalised,
          tool: toolName,
          agentId: this.options.agentId,
        });
      }
    } else {
      this.options.onAllow?.({
        tool: toolName,
        params,
        decision: normalised.decision as 'allow' | 'monitor',
        riskScore: normalised.riskScore,
        matchedRuleId: normalised.matchedRuleId,
      });
    }

    return { blocked, ...normalised };
  }

  /**
   * Wrap a single tool function with AgentGuard policy checks.
   *
   * @param toolName  Name of the tool (for policy evaluation)
   * @param fn        The original tool function
   * @returns         A wrapped function that evaluates policy before executing
   */
  wrapTool<T extends ToolFunction>(toolName: string, fn: T): T {
    const self = this;
    const wrapped = async function (this: unknown, ...args: unknown[]): Promise<unknown> {
      // Extract params: first arg if it's an object, otherwise wrap all args
      const params: Record<string, unknown> =
        args.length === 1 && typeof args[0] === 'object' && args[0] !== null
          ? (args[0] as Record<string, unknown>)
          : { args };

      const result = await self.evaluate(toolName, params);
      if (result.blocked) {
        return undefined;
      }

      return fn.apply(this, args);
    };

    // Preserve function name and properties
    Object.defineProperty(wrapped, 'name', { value: fn.name || toolName });
    return wrapped as unknown as T;
  }

  /**
   * Patch an AutoGen ConversableAgent to route all tool executions
   * through AgentGuard.
   *
   * Supports multiple AutoGen versions by detecting available tool
   * registration patterns (duck-typed).
   *
   * @param agent  An AutoGen ConversableAgent (or structurally compatible object)
   */
  patchAgent(agent: ConversableAgentLike): void {
    // Strategy 1: Patch the tool execution method (v0.4+)
    if (typeof agent.execute_function === 'function') {
      const originalExecute = agent.execute_function.bind(agent);
      agent.execute_function = async (
        toolName: string,
        args: Record<string, unknown>,
      ): Promise<unknown> => {
        const result = await this.evaluate(toolName, args);
        if (result.blocked) {
          return { error: `AgentGuard blocked tool "${toolName}": ${result.reason ?? 'policy violation'}` };
        }
        return originalExecute(toolName, args);
      };
      return;
    }

    // Strategy 2: Patch execute_tool (v0.2+)
    if (typeof agent.execute_tool === 'function') {
      const originalExecute = agent.execute_tool.bind(agent);
      agent.execute_tool = async (
        toolName: string,
        args: Record<string, unknown>,
      ): Promise<unknown> => {
        const result = await this.evaluate(toolName, args);
        if (result.blocked) {
          return { error: `AgentGuard blocked tool "${toolName}": ${result.reason ?? 'policy violation'}` };
        }
        return originalExecute(toolName, args);
      };
      return;
    }

    // Strategy 3: Wrap individual tool functions in _tools or registered_tools
    const toolMap = agent._tools ?? agent.registered_tools;
    if (toolMap && typeof toolMap === 'object') {
      for (const [name, fn] of Object.entries(toolMap)) {
        if (typeof fn === 'function') {
          toolMap[name] = this.wrapTool(name, fn);
        }
      }
      return;
    }

    throw new Error(
      'AutoGenToolGuard: Could not detect tool registration on this agent. ' +
      'Expected execute_function, execute_tool, _tools, or registered_tools.',
    );
  }
}

// ─── Factory function ─────────────────────────────────────────────────────────

/**
 * Convenience guard object for wrapping AutoGen tool functions.
 *
 * Returns an object with `wrapTool` and `evaluate` methods that can be
 * used without instantiating the full class.
 *
 * @example
 * ```typescript
 * import { createAutoGenGuard } from '@the-bot-club/agentguard';
 *
 * const guard = createAutoGenGuard({ apiKey: 'ag_...' });
 *
 * // Wrap individual tools
 * const guardedReadFile = guard.wrapTool('read_file', readFile);
 * const guardedSendEmail = guard.wrapTool('send_email', sendEmail);
 *
 * // Or evaluate manually
 * const result = await guard.evaluate('exec', { command: 'ls' });
 * if (!result.blocked) {
 *   await exec({ command: 'ls' });
 * }
 *
 * // Or patch an entire agent
 * guard.patchAgent(myConversableAgent);
 * ```
 *
 * @param options  Guard configuration
 * @returns        Guard object with wrapTool, evaluate, and patchAgent methods
 */
export function createAutoGenGuard(options: AutoGenGuardOptions): AutoGenToolGuard {
  return new AutoGenToolGuard(options);
}

File: packages/python/agentguard/integrations/autogen.py
"""AgentGuard — AutoGen Integration (Python).

Production-quality integration for Microsoft's AutoGen framework (v0.2+ and
v0.4+). Provides two integration styles:

1. **Decorator-based** — wrap individual tool functions:

    .. code-block:: python

        from agentguard.integrations.autogen import AutoGenGuard

        guard = AutoGenGuard(api_key="ag_...")

        @guard.guard_tool
        def read_file(path: str) -> str:
            return open(path).read()

2. **Agent patching** — guard all tools on an existing ConversableAgent:

    .. code-block:: python

        from agentguard.integrations.autogen import create_guarded_agent

        guarded = create_guarded_agent(agent, api_key="ag_...")

Both approaches use duck-typing — no hard dependency on ``autogen`` or
``pyautogen``. Works with any structurally compatible agent or tool function.
"""
from __future__ import annotations

import functools
import inspect
import json
import logging
from typing import (
    Any,
    Callable,
    Dict,
    List,
    Optional,
    TypeVar,
    Union,
    overload,
)

from ..client import AgentGuard
from .errors import AgentGuardBlockError

logger = logging.getLogger("agentguard.integrations.autogen")

F = TypeVar("F", bound=Callable[..., Any])


# ─── Result types ──────────────────────────────────────────────────────────────


class GuardResult:
    """Result of a tool call policy evaluation."""

    __slots__ = (
        "blocked",
        "decision",
        "risk_score",
        "reason",
        "suggestion",
        "alternatives",
        "matched_rule_id",
    )

    def __init__(self, raw: Dict[str, Any], *, blocked: bool) -> None:
        self.blocked: bool = blocked
        self.decision: str = raw.get("result", raw.get("decision", "allow"))
        self.risk_score: int = raw.get("riskScore", raw.get("risk_score", 0))
        self.reason: Optional[str] = raw.get("reason")
        self.suggestion: Optional[str] = raw.get("suggestion")
        self.alternatives: List[str] = raw.get("alternatives", [])
        self.matched_rule_id: Optional[str] = raw.get(
            "matchedRuleId", raw.get("matched_rule_id")
        )

    def to_dict(self) -> Dict[str, Any]:
        return {
            "blocked": self.blocked,
            "decision": self.decision,
            "risk_score": self.risk_score,
            "reason": self.reason,
            "suggestion": self.suggestion,
            "alternatives": self.alternatives,
            "matched_rule_id": self.matched_rule_id,
        }


# ─── AutoGenGuard ──────────────────────────────────────────────────────────────


class AutoGenGuard:
    """Guard for AutoGen tool functions.

    Provides a ``guard_tool`` decorator that wraps tool functions with
    AgentGuard policy checks, and an ``evaluate`` method for manual
    pre-execution checks.

    Args:
        api_key:        AgentGuard API key (``ag_...``)
        base_url:       Override the AgentGuard API base URL
        agent_id:       Optional agent ID for scoped evaluations
        throw_on_block: When True, raise ``AgentGuardBlockError`` on block.
                        When False (default), call ``on_block`` and return None.
        on_block:       Callback invoked when a tool call is blocked.
                        Receives a ``GuardResult`` instance.
        on_allow:       Callback invoked when a tool call is allowed.
                        Receives a ``GuardResult`` instance.

    Example::

        guard = AutoGenGuard(api_key="ag_...")

        @guard.guard_tool
        def send_email(to: str, subject: str, body: str) -> str:
            # Only executes if AgentGuard allows it
            ...

        # Use with AutoGen's register_for_execution pattern:
        @user_proxy.register_for_execution()
        @assistant.register_for_llm(description="Send an email")
        @guard.guard_tool
        def send_email(to: str, subject: str, body: str) -> str:
            ...
    """

    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.agentguard.tech",
        agent_id: Optional[str] = None,
        throw_on_block: bool = False,
        on_block: Optional[Callable[[GuardResult], None]] = None,
        on_allow: Optional[Callable[[GuardResult], None]] = None,
    ) -> None:
        self._agent_id = agent_id
        self._throw_on_block = throw_on_block
        self._on_block = on_block
        self._on_allow = on_allow
        self._guard = AgentGuard(api_key=api_key, base_url=base_url)

    def evaluate(
        self,
        tool_name: str,
        params: Optional[Dict[str, Any]] = None,
    ) -> GuardResult:
        """Evaluate a tool call against the policy engine.

        Args:
            tool_name: Name of the tool to evaluate
            params:    Tool arguments as a dict

        Returns:
            A ``GuardResult`` with the policy decision.

        Raises:
            AgentGuardBlockError: If ``throw_on_block=True`` and the tool is blocked.
        """
        raw = self._guard.evaluate(tool=tool_name, params=params or {})
        decision = raw.get("result", raw.get("decision", "allow"))
        blocked = decision in ("block", "require_approval")

        result = GuardResult(raw, blocked=blocked)

        if blocked:
            if self._on_block is not None:
                self._on_block(result)

            if self._throw_on_block:
                raise AgentGuardBlockError(
                    {
                        **raw,
                        "tool": tool_name,
                        "decision": decision,
                        "agent_id": self._agent_id,
                    }
                )
        else:
            if self._on_allow is not None:
                self._on_allow(result)

        return result

    @overload
    def guard_tool(self, func: F) -> F: ...

    @overload
    def guard_tool(
        self, *, name: Optional[str] = None
    ) -> Callable[[F], F]: ...

    def guard_tool(
        self,
        func: Optional[F] = None,
        *,
        name: Optional[str] = None,
    ) -> Union[F, Callable[[F], F]]:
        """Decorator that wraps a tool function with AgentGuard policy checks.

        Can be used with or without arguments::

            @guard.guard_tool
            def my_tool(x: int) -> int: ...

            @guard.guard_tool(name="custom_tool_name")
            def my_tool(x: int) -> int: ...

        When the tool is blocked:
        - If ``throw_on_block=True``: raises ``AgentGuardBlockError``
        - If ``throw_on_block=False`` (default): returns ``None`` and calls
          ``on_block`` callback if provided

        Args:
            func: The tool function to wrap (when used without parentheses)
            name: Override the tool name used for policy evaluation.
                  Defaults to the function's ``__name__``.

        Returns:
            The wrapped function (same signature as the original).
        """
        def decorator(fn: F) -> F:
            tool_name = name or fn.__name__

            if inspect.iscoroutinefunction(fn):
                @functools.wraps(fn)
                async def async_wrapper(*args: Any, **kwargs: Any) -> Any:
                    params = _extract_params(fn, args, kwargs)
                    result = self.evaluate(tool_name, params)
                    if result.blocked:
                        return None
                    return await fn(*args, **kwargs)

                return async_wrapper  # type: ignore[return-value]
            else:
                @functools.wraps(fn)
                def sync_wrapper(*args: Any, **kwargs: Any) -> Any:
                    params = _extract_params(fn, args, kwargs)
                    result = self.evaluate(tool_name, params)
                    if result.blocked:
                        return None
                    return fn(*args, **kwargs)

                return sync_wrapper  # type: ignore[return-value]

        if func is not None:
            # Used as @guard.guard_tool (no parentheses)
            return decorator(func)

        # Used as @guard.guard_tool(name="...") (with parentheses)
        return decorator

    def wrap_tool(
        self,
        tool_name: str,
        func: F,
    ) -> F:
        """Wrap a tool function with policy checks (non-decorator style).

        Args:
            tool_name: Name of the tool for policy evaluation
            func:      The original tool function

        Returns:
            A wrapped function that evaluates policy before executing.

        Example::

            guarded_read = guard.wrap_tool("read_file", read_file)
            result = guarded_read(path="/etc/passwd")
        """
        return self.guard_tool(func, name=tool_name)

    def patch_agent(self, agent: Any) -> None:
        """Patch an AutoGen ConversableAgent to guard all registered tools.

        Supports multiple AutoGen versions by detecting available tool
        registration patterns (duck-typed):

        - ``_tools`` dict (v0.2 ConversableAgent)
        - ``registered_tools`` dict
        - ``_function_map`` dict (v0.4+)

        Args:
            agent: An AutoGen ConversableAgent or structurally compatible object.

        Raises:
            TypeError: If no tool registration mechanism is detected.

        Example::

            from autogen import ConversableAgent

            agent = ConversableAgent(name="assistant", ...)
            guard.patch_agent(agent)
            # All tool calls now go through AgentGuard
        """
        patched = False

        # v0.4+: _function_map
        func_map = getattr(agent, "_function_map", None)
        if isinstance(func_map, dict) and func_map:
            for fn_name, fn in list(func_map.items()):
                if callable(fn):
                    func_map[fn_name] = self.wrap_tool(fn_name, fn)
                    patched = True

        # v0.2+: _tools
        tools = getattr(agent, "_tools", None)
        if isinstance(tools, dict) and tools:
            for fn_name, fn in list(tools.items()):
                if callable(fn):
                    tools[fn_name] = self.wrap_tool(fn_name, fn)
                    patched = True

        # Alternative: registered_tools
        registered = getattr(agent, "registered_tools", None)
        if isinstance(registered, dict) and registered:
            for fn_name, fn in list(registered.items()):
                if callable(fn):
                    registered[fn_name] = self.wrap_tool(fn_name, fn)
                    patched = True

        if not patched:
            raise TypeError(
                "AutoGenGuard.patch_agent: Could not detect tool registration "
                "on this agent. Expected _function_map, _tools, or "
                "registered_tools attribute with callable values."
            )

        logger.info("Patched agent with AgentGuard tool guards")


# ─── Convenience factory ───────────────────────────────────────────────────────


def create_guarded_agent(
    agent: Any,
    api_key: str,
    base_url: str = "https://api.agentguard.tech",
    agent_id: Optional[str] = None,
    throw_on_block: bool = False,
    on_block: Optional[Callable[[GuardResult], None]] = None,
    on_allow: Optional[Callable[[GuardResult], None]] = None,
) -> Any:
    """Patch an existing AutoGen agent with AgentGuard policy enforcement.

    Convenience function that creates an ``AutoGenGuard`` and patches the
    agent in one call. Returns the same agent (mutated in-place).

    Args:
        agent:          An AutoGen ConversableAgent or compatible object
        api_key:        AgentGuard API key (``ag_...``)
        base_url:       Override the AgentGuard API base URL
        agent_id:       Optional agent ID for scoped evaluations
        throw_on_block: Raise on block (default False — zero-throw)
        on_block:       Callback for blocked tool calls
        on_allow:       Callback for allowed tool calls

    Returns:
        The same agent, with all tool functions wrapped by AgentGuard.

    Example::

        from autogen import ConversableAgent
        from agentguard.integrations.autogen import create_guarded_agent

        agent = ConversableAgent(name="assistant", ...)
        agent.register_for_execution()(my_tool)

        create_guarded_agent(agent, api_key="ag_...")
        # agent's tools are now guarded
    """
    guard = AutoGenGuard(
        api_key=api_key,
        base_url=base_url,
        agent_id=agent_id,
        throw_on_block=throw_on_block,
        on_block=on_block,
        on_allow=on_allow,
    )
    guard.patch_agent(agent)
    return agent


# ─── Internal helpers ──────────────────────────────────────────────────────────


def _extract_params(
    func: Callable[..., Any],
    args: tuple,
    kwargs: Dict[str, Any],
) -> Dict[str, Any]:
    """Extract a params dict from function call arguments.

    Binds positional and keyword args to the function's parameter names
    so that policy evaluation sees named parameters regardless of how
    the function was called.
    """
    params: Dict[str, Any] = dict(kwargs)

    try:
        sig = inspect.signature(func)
        bound = sig.bind_partial(*args, **kwargs)
        bound.apply_defaults()
        params = dict(bound.arguments)
    except (ValueError, TypeError):
        # Fallback: merge positional args as indexed keys
        if args:
            params["_positional_args"] = list(args)

    # Remove 'self'/'cls' if present (method calls)
    params.pop("self", None)
    params.pop("cls", None)

    # Ensure all values are JSON-serializable for the API
    sanitised: Dict[str, Any] = {}
    for k, v in params.items():
        try:
            json.dumps(v)
            sanitised[k] = v
        except (TypeError, ValueError):
            sanitised[k] = str(v)

    return sanitised
